packages/flutter_tools/gradle/src/main/groovy/flutter.groovy

```diff
@@ -851,17 +864,53 @@ class FlutterPlugin implements Plugin<Project> {
                     if (getCompileSdkFromProject(pluginProject).isInteger()) {
                         pluginCompileSdkVersion = getCompileSdkFromProject(pluginProject) as int
                     }
+
                     maxPluginCompileSdkVersion = Math.max(pluginCompileSdkVersion, maxPluginCompileSdkVersion)
+                    if (pluginCompileSdkVersion > projectCompileSdkVersion) {
+                        pluginsWithHigherSdkVersion.add(new Tuple(pluginProject.name, pluginCompileSdkVersion))
+                    }
+
                     String pluginNdkVersion = pluginProject.android.ndkVersion ?: ndkVersionIfUnspecified
                     maxPluginNdkVersion = mostRecentSemanticVersion(pluginNdkVersion, maxPluginNdkVersion)
+                    if (pluginNdkVersion != projectNdkVersion) {
+                        pluginsWithDifferentNdkVersion.add(new Tuple(pluginProject.name, pluginNdkVersion))
+                    }
 
                     numProcessedPlugins--
                     if (numProcessedPlugins == 0) {
                         if (maxPluginCompileSdkVersion > projectCompileSdkVersion) {
-                            project.logger.error("One or more plugins require a higher Android SDK version.\nFix this issue by adding the following to ${project.projectDir}${File.separator}build.gradle:\nandroid {\n  compileSdkVersion ${maxPluginCompileSdkVersion}\n  ...\n}\n")
+                            project.logger.error("Your project is configured to compile against Android SDK $projectCompileSdkVersion, but the following plugin(s) require to be compiled against a higher Android SDK version:")
```

Comment:
A source I found regarding wording:
- https://stackoverflow.com/q/65278287

```suggestion
                            project.logger.error("Your project is configured to compile using Android SDK $projectCompileSdkVersion, but the following plugin(s) require a higher Android SDK version:")
```

Reply:
Thank you!

Reply:
On a second thought, I'm not sure if "using" is always a right replacement. For example:

_Fix this issue by compiling using the highest Android SDK version (they are backward compatible)._

I don't think it sounds good – "compiling using _something_" sounds as if this _something_ was a compiler, build tool, or similar thing that does work.

But in this case, Android SDK are merely header files.

Reply:
I have pretty much zero knowledge of the Android SDK (I had no idea that it's more of a reference than something actively involved in the compilation), so I really appreciate your explanation.

My concern is that beginners won't understand this colloquialism, and they rely on error messages more than others do.

Outside of the software industry, it's much more common to say something like "My husband fixed the printer using the owner's manual" rather than "My husband fixed the printer against the owner's manual"—the manual was still "used" as a reference, even if it wasn't actively doing any of the work.

And even within the software industry, the word "using" appears to be more common: you configure the SDK version with the tag [`<uses-sdk>`](https://developer.android.com/guide/topics/manifest/uses-sdk-element) rather than `<against-sdk>`.

<br>

But this is all extremely pedantic—I'd still be very happy to see this PR merged as-is.

Reply:
Thank you for this extensive explanation and discussion. However this time I choose to stand by wording! :P

I think "compile against" is good enough – it stresses that there's no direct dependency (as e.g. Gradle's `implementation`), but rather the SDK.

> I have pretty much zero knowledge of the Android SDK (I had no idea that it's more of a reference than something actively involved in the compilation)

Yes, it's confusing and not well explained. But `compileSdk` is actually a good name - it's the SDK version that will be used when compiling your app - and nothing more. It will not be present in the final APK's AndroidManifest (only `minSdk` and `targetSdk` are).

The `<uses-sdk`> AndroidManifest attribute you mention has 3 values:
- `minSdkVersion` - this comes from `minSdk` in `build.gradle`
- `targetSdkVersion` - this comes from `targetSdk` in `build.gradle`
- `maxSdkVersion` - The docs also say it should almost never be used.

So `compileSdk` is only used at compilation-time.

Reply:
FWIW @bartekpacia I like the phrasing you selected.